File: src/akgentic/core/user_proxy.py
"""UserProxy agent for human-in-the-loop workflows.

Migrated from akgentic-framework/libs/akgentic/akgentic/core/user_proxy.py
with v2 additions: ResultMessage routing and receiveMsg_UserMessage handler.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from akgentic.core.agent import Akgent
from akgentic.core.agent_config import BaseConfig
from akgentic.core.agent_state import BaseState
from akgentic.core.messages.message import Message, ResultMessage, UserMessage

if TYPE_CHECKING:
    from akgentic.core.actor_address import ActorAddress

logger = logging.getLogger(__name__)


class UserProxy(Akgent[BaseConfig, BaseState]):
    """Agent that enables human-in-the-loop workflows.

    UserProxy acts as a bridge between human users and the agent system.
    It receives messages from agents requesting human input, presents them
    to humans, and routes human responses back to the requesting agents.

    Usage:
        >>> from akgentic.core import ActorSystem, UserProxy, BaseConfig
        >>> system = ActorSystem()
        >>> config = BaseConfig(name="human", role="UserProxy")
        >>> proxy_addr = system.createActor(UserProxy, config=config)

        # Agent requests human input
        >>> worker.send(proxy_addr, UserMessage(content="What should I do next?"))

        # Human provides input via process_human_input
        >>> proxy_ref.proxy().process_human_input("Continue with option A", user_msg).get()

    Phase 3 Extension:
        Subclass UserProxy to integrate with UI systems:
        >>> class WebSocketUserProxy(UserProxy):
        ...     def receiveMsg_UserMessage(self, message: UserMessage, sender: ActorAddress):
        ...         self.websocket.send(message.content, reply_to=message)
    """

    def process_human_input(self, content: str, message: Message) -> None:
        """Process human input and route it back to the requesting agent.

        Called externally (e.g., from a UI layer) to inject human responses
        into the agent workflow. The default implementation creates a
        ResultMessage with the human's response and sends it to the original
        sender.

        Override this method in subclasses to customize response handling,
        validation, or routing logic for specific workflow requirements.

        Args:
            content: The human's text response.
            message: The original message from the requesting agent.
                     message.sender is used as the reply-to address.

        Example:
            >>> user_proxy.process_human_input("Approve the plan", original_msg)
        """
        logger.info(f"Received human input: {content}, at destination of: {message.sender}")
        response = ResultMessage(content=content)
        self.send(message.sender, response)

    def receiveMsg_UserMessage(  # noqa: N802
        self, message: UserMessage, sender: ActorAddress
    ) -> None:
        """Handle UserMessage from agents requesting human input.

        Default implementation logs the message. Override in subclasses
        to integrate with specific UI systems (WebSocket, REST API, CLI, etc.).

        Args:
            message: The UserMessage requesting human input.
            sender: The ActorAddress of the requesting agent.

        Example override:
            >>> class CLIUserProxy(UserProxy):
            ...     def receiveMsg_UserMessage(self, message: UserMessage, sender: ActorAddress):
            ...         response = input(f"Agent asks: {message.content}\\n> ")
            ...         self.process_human_input(response, message)
        """
        logger.info(
            f"UserProxy received message from {sender}: "
            f"{message.content if hasattr(message, 'content') else message}"
        )

File: lib/akgentic/user_proxy.ex
defmodule Akgentic.UserProxy do
  @moduledoc """
  UserProxy agent for human-in-the-loop workflows.

  UserProxy acts as a bridge between human users and the agent system.
  It receives signals from agents requesting human input, presents them
  to humans, and routes human responses back to the requesting agents.

  Maps from Python's `UserProxy(Akgent[BaseConfig, BaseState])` class.

  ## Python → Elixir Mapping

  | Python `UserProxy`                     | Elixir `Akgentic.UserProxy`          |
  |----------------------------------------|--------------------------------------|
  | `process_human_input(content, msg)`    | `process_human_input/3`              |
  | `receiveMsg_UserMessage`               | Signal handler via GenServer         |

  ## Examples

      {:ok, proxy} = Akgentic.UserProxy.start_link(name: "human")

      # Agent sends a question to the proxy
      Akgentic.UserProxy.receive_message(proxy, signal)

      # Human provides input
      Akgentic.UserProxy.process_human_input(proxy, "Continue with option A", reply_to: sender_pid)

      # Get pending messages
      pending = Akgentic.UserProxy.get_pending_messages(proxy)
  """

  use GenServer

  require Logger

  defstruct [
    :name,
    pending_messages: [],
    message_handler: nil
  ]

  @type t :: %__MODULE__{
          name: String.t(),
          pending_messages: [Akgentic.Signal.t()],
          message_handler: (Akgentic.Signal.t() -> :ok) | nil
        }

  # --- Public API ---

  @doc """
  Start the UserProxy process.

  ## Options

    * `:name` - Proxy name (default: "user_proxy")
    * `:server_name` - Optional GenServer name for registration
    * `:message_handler` - Optional callback for incoming messages
  """
  @spec start_link(keyword()) :: GenServer.on_start()
  def start_link(opts \\ []) do
    name = Keyword.get(opts, :name, "user_proxy")
    server_name = Keyword.get(opts, :server_name, nil)
    handler = Keyword.get(opts, :message_handler, nil)

    gen_opts = if server_name, do: [name: server_name], else: []

    state = %__MODULE__{
      name: name,
      message_handler: handler
    }

    GenServer.start_link(__MODULE__, state, gen_opts)
  end

  @doc """
  Receive a message/signal from an agent requesting human input.

  The signal is stored in the pending queue and optionally forwarded
  to the message_handler callback.
  """
  @spec receive_message(GenServer.server(), Akgentic.Signal.t()) :: :ok
  def receive_message(server, signal) do
    GenServer.cast(server, {:receive_message, signal})
  end

  @doc """
  Process human input and route it back to the requesting agent.

  ## Options

    * `:reply_to` - PID of the agent to send the response to
    * `:source` - Signal source (default: "/user_proxy")
  """
  @spec process_human_input(GenServer.server(), String.t(), keyword()) :: :ok
  def process_human_input(server, content, opts \\ []) do
    GenServer.cast(server, {:process_human_input, content, opts})
  end

  @doc """
  Get all pending messages waiting for human input.
  """
  @spec get_pending_messages(GenServer.server()) :: [Akgentic.Signal.t()]
  def get_pending_messages(server) do
    GenServer.call(server, :get_pending_messages)
  end

  @doc """
  Clear all pending messages.
  """
  @spec clear_pending(GenServer.server()) :: :ok
  def clear_pending(server) do
    GenServer.cast(server, :clear_pending)
  end

  # --- GenServer Callbacks ---

  @impl true
  def init(state) do
    {:ok, state}
  end

  @impl true
  def handle_cast({:receive_message, signal}, state) do
    Logger.info(
      "UserProxy received message from #{signal.source}: " <>
        "#{inspect(get_in(signal.data, [:content]) || signal.data)}"
    )

    # Call handler if configured
    if state.message_handler do
      try do
        state.message_handler.(signal)
      rescue
        e -> Logger.error("UserProxy message handler failed: #{inspect(e)}")
      end
    end

    {:noreply, %{state | pending_messages: state.pending_messages ++ [signal]}}
  end

  @impl true
  def handle_cast({:process_human_input, content, opts}, state) do
    reply_to = Keyword.get(opts, :reply_to, nil)
    source = Keyword.get(opts, :source, "/user_proxy/#{state.name}")

    Logger.info("Received human input: #{content}, at destination of: #{inspect(reply_to)}")

    if reply_to && is_pid(reply_to) && Process.alive?(reply_to) do
      response = Akgentic.Messages.result_message(content, source: source)

      send(reply_to, {:signal, response})
    end

    {:noreply, state}
  end

  @impl true
  def handle_cast(:clear_pending, state) do
    {:noreply, %{state | pending_messages: []}}
  end

  @impl true
  def handle_call(:get_pending_messages, _from, state) do
    {:reply, state.pending_messages, state}
  end
end
